lib/AST/Type.cpp

```diff
@@ -3532,6 +3566,35 @@ case TypeKind::Id:
 
     return BoundGenericType::get(bound->getDecl(), substParentTy, substArgs);
   }
+      
+  case TypeKind::OpaqueTypeArchetype: {
+    auto opaque = cast<OpaqueTypeArchetypeType>(base);
+    
+    if (opaque->getSubstitutions().empty())
+      return *this;
+    
+    SmallVector<Type, 4> newSubs;
+    bool anyChanged = false;
+    for (auto replacement : opaque->getSubstitutions().getReplacementTypes()) {
+      Type newReplacement = replacement.transformRec(fn);
+      if (!newReplacement)
+        return Type();
+      newSubs.push_back(newReplacement);
+      if (replacement.getPointer() != newReplacement.getPointer())
+        anyChanged = true;
+    }
+    
+    if (!anyChanged)
+      return *this;
+    
+    // TODO: Update protocol conformances?
+    auto newSubMap =
+      SubstitutionMap::get(opaque->getOpaqueDecl()->getGenericSignature(),
+                           newSubs,
+                           opaque->getSubstitutions().getConformances());
```

Comment:
If the transform is doing anything interesting, passing along conformances will be wrong. It's probably safest to use `SubstitutionMap::get(the-generic-signature, a-function-that-uses-the-replaced-types, LookupConformanceInModule(module-of-the-opaque-decl)`.

Reply:
Thanks, I was aware of that problem but wasn't sure how to force it to re-lookup the conformances. It seems like `transformRec` really ought to be working with something akin to the old `Substitution` record instead of types, though.